microsoft/react-native-macos
# Add Hermes engine as option for macOS

Related project board https://github.com/microsoft/react-native-macos/projects/4

#### Please select one of the following


## Summary

This adds the [Hermes engine](https://hermesengine.dev) as an optional replacement to JavaScriptCore for macOS targets.

* Hermes is enabled by virtue of adding the `React-Core/Hermes` dependency to one’s Podfile.

* The engine is shipped as a prebuilt framework binary, optimised for release.

* It comes with a separate debug symbols bundle (dSYM) for debugging/crash report symbolication.
  
   <img width="1225" alt="hermes-framework-loaded-with-dSYM" src="https://user-images.githubusercontent.com/2320/85873066-ec137480-b7d0-11ea-9418-155541500326.png">

   <img width="1277" alt="availability-of-debug-symbols" src="https://user-images.githubusercontent.com/2320/85874579-23832080-b7d3-11ea-81ee-95d981392797.png">

* Support to generate source-maps when building release builds has been added, both without Hermes enabled:
   
   <img width="1588" alt="source-maps-jsc" src="https://user-images.githubusercontent.com/2320/85873164-0baa9d00-b7d1-11ea-9235-20e35accdecb.png">

   …and with Hermes enabled:

   <img width="1587" alt="source-maps-hermes" src="https://user-images.githubusercontent.com/2320/85873175-0e0cf700-b7d1-11ea-9a84-a76f3b91ec3b.png">

   (Note that file/line is correct, but column not. At the moment I am brushing this off as a minor issue, but something to watch out for, especially when we move to Hermes `master`.)

## Changelog

[macOS] [Added] - Add Hermes as (optional) JS engine.
[macOS/iOS] [Added] - Add ability to generate source-maps when building a release build with Xcode.

## Test Plan

* Update `Podfile` with `use_react_native!(hermes_enabled => true)` to add the dependencies and enable it as runtime engine.
* Uncomment `FORCE_BUNDLING` in bundle build-script phase to test source-maps.
* Verify that attaching a Chrome debugger works.

## TODO

- [x] Finish this PR: review it myself and finish the PR description.
- [x] Rebase on latest RN macOS v0.62.0 branch.
- [ ] The changes required to build the binary are currently still [here in my fork](https://github.com/alloy/hermes/tree/add-binary-podspec-for-macos-v0.4.1). I am reaching out to the upstream team about incorporating those changes upstream and releasing npm packages with the binary from their account, especially going forward.

    This could also be done in a follow-up PR, as long as we do so before we release v0.62.0 (preview); although in that case it might be preferable to move the tarball from my account to this MS repo?

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-macos/pull/473)
